raft/raft.go

```diff
@@ -437,11 +437,16 @@ func (r *raft) maybeSendAppend(to uint64, sendIfEmpty bool) bool {
 	if pr.IsPaused() {
 		return false
 	}
-	m := pb.Message{}
-	m.To = to
+	m := pb.Message{To: to}
 
 	term, errt := r.raftLog.term(pr.Next - 1)
-	ents, erre := r.raftLog.entries(pr.Next, r.maxMsgSize)
+	var ents []pb.Entry
+	var erre error
+	// In a throttled StateReplicate only send empty MsgApp, to ensure progress.
```

Comment:
Also, could you comment on the significance of `sendIfEmpty`? If that is `false`, then if the follower's inflight is full we are definitely returning on line 451. Is that what we want? Feels like we'd really want to get the entries unconditionally, and then, if there are entries, set `sendIfEmpty` to false, and if the follower is paused send an empty append anyway (instead of a "full" append).

Right now I'm left wondering why we have a caller that passes `sendIfEmpty==true` in when we need it. I know now, after looking at the handling of `MsgHeartbeatResp`, that we will in fact pass `sendIfEmpty==true`.

Having looked into all of this, the code here checks out - we want to send the empty append resulting from being full only if the caller requested empty appends to be sent anyway. This is the case on heartbeats (and in some other cases, for example whenever the commit index changes). But try to package all of this lifecycle into a comment somewhere.

Reply:
> Could add here:

Done.

> Also, could you check that if Inflights is full, and the Log{Term,Index} in the empty MsgApp that reaches the follower actually matches the end of their log, the follower still responds with an MsgAppResp? If it didn't then we'd get stuck anyway.

Verified, seems ok. See the data-driven test.

> I'd also appreciate a TestInteraction test for a follower that gets these empty appends, because it essentially serves as an end-to-end test.

Done. See the first commit introducing the data-driven test. Note the TODO in it which flags the issue that this PR solves. Note the change in the data-driven test after the fix.

> Also, isn't there a (pre-existing) bug here?

Is this addressable independently? Let's perhaps move it to a separate issue/PR to keep this one self-contained.

> But try to package all of this lifecycle into a comment somewhere.

Not yet done, will do tomorrow.

Reply:
All done, except the `sendIfEmpty` commenting. I think that `maybeSendAppend()` method comment speaks about it quite ok, wasn't sure what your concern was. Do you have a specific place/comment in mind?